Implement GPS for Canon EOS R.

gkoh opened the pull request:
In collaboration with @Jerroder, implement the GPS sync protocol for Canon EOS R.

gkoh commented:
Thanks goes to @Jerroder for the HCI logs and initial protocol work that made this possible.

Jerroder commented:
You can now remove it from the "things to do" in the README 😉

Oaksen commented:
I started look at the EOS R geo service too after getting annoyed at the official app, then found this after :)

I've just tested this and unfortunately I don't think it'll work as-is. The pairing method you've got for the EOSR is using the wireless remote service, and the camera doesn't advertise the geo service in that mode. I think it needs to be paired as a smartphone using the service you've got for the M6.

gkoh commented:
> I started look at the EOS R geo service too after getting annoyed at the official app, then found this after :)
> 
> I've just tested this and unfortunately I don't think it'll work as-is. The pairing method you've got for the EOSR is using the wireless remote service, and the camera doesn't advertise the geo service in that mode. I think it needs to be paired as a smartphone using the service you've got for the M6.

Hello @Oaksen, thank you for the effort.
Unfortunately I'm coding completely blind here, no EOSR as a test target :frowning_face:
Thus I rely on the kindness of strangers.

What you mention sounds similar to the Nikon, when pairing to a wireless remote accessory a specific service UUID appears.

The EOSR implementation was changed to use the remote service primarily to get focus working.
Does the official app support a focus operation with the EOSR cameras?

gkoh commented:
Looking around, I don't think the official app supports focus.
However, I also realise that this PR is useless as it stands, it will never work (thanks @Oaksen).

Thus, I will refactor the GPS code into the M6 implementation (and probably rename them (M6->smart, R->remote).
I will then move the content of this PR from 'remote' -> 'smart'.
This will allow the user to choose to connect either:
- remote mode with shutter + focus, but no GPS
- smart device mode with shutter + GPS, but no focus

gkoh commented:
I have refactored the GPS support into the smart mode used by the M6.
And I've cut a test release here:
https://github.com/gkoh/furble/releases/tag/v3.3.0-rc1

NB: I changed the `match()` code to use the primary service UUID rather than manufacturer ID and advertising data internals.
This may not work at all ...

hijae commented:
I just tested it and it doesn't work.
Using the phone app, there is no focus function.

gkoh commented:
> Then I'd say it sounds about right unfortunately 😕
> 
> The official Canon app doesn't have a focus button, and using the same services as they do I don't discover a service UUID starting with `0x00050000`. My shutter characteristic is `00030031-0000-1000-0000-d8492fffa821` and it notifies when autofocus fails (because triggering the shutter starts autofocus), but I can't initiate only autofocus (as documented in https://github.com/3bl3gamer/canon-bluetooth-control).

😢
Thanks for confirming, it is what it is.

gkoh commented:
@Oaksen I have merged in your changes and built a test release here:
https://github.com/gkoh/furble/releases/tag/vTEST-189a

If you're able, a test would be much appreciated.

gkoh commented:
Fastest way to get visibility and testing on this is to release and wait for the screams :slightly_smiling_face:
